### PR TITLE
CBG-321: Create one logging function and external logger support

### PR DESCRIPTION
**Pull request by JRascagneres**

### CODE REVIEW

File: logging.go [hunk 1]
```diff
@@ -1,35 +1,45 @@
...
 	}
 }
 
-func prefixWithReplicationId(lr LoggingReplication, raw string) string {
-
-	replicationId := lr.Parameters.ReplicationId
-
-	if replicationId != "" {
-		return strings.Join([]string{"[", replicationId, "] ", raw}, "")
+var defaultLogFn = func(level clog.LogLevel, format string, args ...interface{}) {
```
Comment (bbrks) [inline logging.go]: I don't think we should make this a variable, it isn't going to change. How about a normal function?

File: logging.go [hunk 2]
```diff
@@ -1,35 +1,45 @@
...
-	clog.To(key, prefixWithReplicationId(lr, format), args...)
+func (r *ContinuousReplication) log(level clog.LogLevel, format string, args ...interface{}) {
+	log(r.Parameters, level, format, args)
 }
 
-func (lr LoggingReplication) Warn(args ...interface{}) {
-	if clog.Level <= clog.LevelWarning {
-		clog.Warn(prefixWithReplicationId(lr, fmt.Sprint(args...)))
+func log(param ReplicationParameters, level clog.LogLevel, format string, args ...interface{}) {
+	if param.LogFn == nil {
```
Comment (bbrks) [inline logging.go]: Doing this check on each invocation of log seems a bit much?

Can we not rely on setting the default in `NewReplication()`?

File: logging.go [hunk 3]
```diff
@@ -1,35 +1,45 @@
...
+		param.LogFn = defaultLogFn
+	}
+	if param.ReplicationId == "" {
+		param.LogFn(level, format, args...)
+	} else {
+		param.LogFn(level, "["+param.ReplicationId+"]"+format, args...)
```
Comment (bbrks) [inline logging.go]: You might need a space here after "]" to prevent formatting issues?

File: logging.go [hunk 4]
```diff
@@ -1,35 +1,45 @@
...
 }
 
-func (lr LoggingReplication) Warn(args ...interface{}) {
-	if clog.Level <= clog.LevelWarning {
-		clog.Warn(prefixWithReplicationId(lr, fmt.Sprint(args...)))
+func log(param ReplicationParameters, level clog.LogLevel, format string, args ...interface{}) {
+	if param.LogFn == nil {
+		param.LogFn = defaultLogFn
+	}
+	if param.ReplicationId == "" {
```
Comment (bbrks) [inline logging.go]: Could probably simplify this part a little

```
	if param.ReplicationId != "" {
		format = "["+param.ReplicationId+"] "+format
	}
	param.LogFn(level, format, args...)
```

File: logging.go (hunk 2 above)
Comment (JRascagneres) [reply logging.go]: This doesn't work in the tests as the 'constructor' isn't always ran first. This check was for test logging... Perhaps there may be another way around.

Comment (JRascagneres) [reply logging.go]: I guess moving the work to the test side may be better my TestMain seems to be too much for it?

File: cli/cli_parser_test.go [hunk 5]
```diff
@@ -4,10 +4,10 @@ import (
 	"strings"
 	"testing"
 
+	"fmt"
```
Comment (bbrks) [inline cli/cli_parser_test.go]: group with stdlib imports above

File: continuous_replication.go [hunk 6]
```diff
@@ -116,40 +120,40 @@ func (r *ContinuousReplication) GetStats() *ReplicationStats {
...
+	r.log(clog.LevelNormal, "ContinuousReplication.processEvents()")
 
 	// nil out the EventChan after the event loop has finished.
 	defer func() { r.EventChan = nil }()
 
 	defer close(r.NotificationChan) // No more notifications
 
 	for state := stateFnCatchingUp; state != nil; {
-		r.LogTo("Replicate", "continuous repl state: %v", state)
+		r.log(clog.LevelNormal, "Replicate", "continuous repl state: %v", state)
```
Comment (bbrks) [inline continuous_replication.go]: I think "Replicate" was accidentally left in here

File: continuous_replication.go [hunk 7]
```diff
@@ -116,40 +120,40 @@ func (r *ContinuousReplication) GetStats() *ReplicationStats {
...
 		state = state(r)
-		r.LogTo("Replicate", "continuous repl new state: %v", state)
+		r.log(clog.LevelNormal, "Replicate", "continuous repl new state: %v", state)
```
Comment (bbrks) [inline continuous_replication.go]: I think "Replicate" was accidentally left in here

File: continuous_replication.go [hunk 8]
```diff
@@ -116,40 +120,40 @@ func (r *ContinuousReplication) GetStats() *ReplicationStats {
...
 	}
-	r.LogTo("Replicate", "continuous repl processEvents() is done")
+	r.log(clog.LevelNormal, "Replicate", "continuous repl processEvents() is done")
```
Comment (bbrks) [inline continuous_replication.go]: I think "Replicate" was accidentally left in here

File: logging.go [hunk 9]
```diff
@@ -1,35 +1,41 @@
...
-		return strings.Join([]string{"[", replicationId, "] ", raw}, "")
+func defaultLogFn(level clog.LogLevel, format string, args ...interface{}) {
+	if clog.GetLevel() <= level {
+		switch level {
+		case clog.LevelDebug:
+			clog.Debugf(format, args...)
+		case clog.LevelNormal:
+			clog.To(logKey, format, args...)
+		case clog.LevelWarning:
+			clog.Debugf(format, args...)
```
Comment (adamcfraser) [inline logging.go]: Should be warn?

**Comment by JRascagneres** [comment]
> Still a bunch of LevelNormal logging in here

I noticed earlier. My Ctrl replace didn't catch them all. Will do shortly

**Pull request merged**